### PR TITLE
[#9595] Add tests for GET /auth

### PR DESCRIPTION
**Pull request by crizyli**
Part of #9595

Add tests for csrf token cookie in `GetAuthInfoActionTest.java`

### CODE REVIEW

File: src/test/java/teammates/test/cases/webapi/GetAuthInfoActionTest.java [hunk 1]
```diff
@@ -126,8 +134,38 @@ protected void testExecute() {
                 Const.ParamsNames.USER_ID, "idOfAnotherInstructor",
         }));
 
-        // TODO test CSRF token cookies
+        ______TS("Test adding CSRF token cookies");
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/webapi/GetAuthInfoActionTest.java]: Testing CSRF token can be a separate `@Test`?

File: src/test/java/teammates/test/cases/webapi/GetAuthInfoActionTest.java [hunk 2]
```diff
@@ -126,8 +134,38 @@ protected void testExecute() {
...
+
+        loginAsInstructor("idOfInstructor1OfCourse1");
+
+        MockHttpServletRequest mockRequest = new MockHttpServletRequest(getRequestMethod(),
+                Const.ResourceURIs.URI_PREFIX + getActionUri());
+        mockRequest.addCookie(new Cookie(Const.CsrfConfig.TOKEN_COOKIE_NAME, "some random csrf token"));
+        MockHttpServletResponse mockResponse = new MockHttpServletResponse();
+
+        a = (GetAuthInfoAction) new ActionFactory().getAction(mockRequest, getRequestMethod(), mockResponse);
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/webapi/GetAuthInfoActionTest.java]: Let not break the abstraction from the base action. You can write another `getAction` in BaseActionTest.

File: src/test/java/teammates/test/cases/webapi/GetAuthInfoActionTest.java [hunk 3]
```diff
@@ -126,8 +134,38 @@ protected void testExecute() {
                 Const.ParamsNames.USER_ID, "idOfAnotherInstructor",
         }));
 
-        // TODO test CSRF token cookies
+        ______TS("Test adding CSRF token cookies");
+
+        loginAsInstructor("idOfInstructor1OfCourse1");
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/webapi/GetAuthInfoActionTest.java]: CSRF token will be sent regardless of login status. So you should have test case when users is not logged-in?

**Comment by crizyli** [comment]
Thanks! @xpdavid I have updated with those issues.

File: src/test/java/teammates/test/cases/webapi/GetAuthInfoActionTest.java [hunk 4]
```diff
@@ -125,9 +130,66 @@ protected void testExecute() {
...
+        ______TS("Test will add CSRF token cookies: No logged in user");
+
+        gaeSimulation.logoutUser();
+
+        MockHttpServletRequest mockRequest = new MockHttpServletRequest(getRequestMethod(),
+                Const.ResourceURIs.URI_PREFIX + getActionUri());
+        mockRequest.addCookie(new Cookie(Const.CsrfConfig.TOKEN_COOKIE_NAME, "some random csrf token"));
+        MockHttpServletResponse mockResponse = new MockHttpServletResponse();
+
+        GetAuthInfoAction a = getAction(mockRequest, getRequestMethod(), mockResponse);
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/webapi/GetAuthInfoActionTest.java]: You can add a new method `getHttpServletRequest()` in `GetAuthInfoAction` to get the mocked request.

File: src/test/java/teammates/test/cases/webapi/GetAuthInfoActionTest.java [hunk 5]
```diff
@@ -125,9 +130,66 @@ protected void testExecute() {
...
+        getJsonResult(a);
+
+        assertEquals(StringHelper.encrypt(mockRequest.getSession().getId()), mockResponse.getCookie().getValue());
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/webapi/GetAuthInfoActionTest.java]: Consider use the method inside `HttpRequestHelper` to get cookie.

File: src/test/java/teammates/test/cases/webapi/GetAuthInfoActionTest.java [hunk 6]
```diff
@@ -125,9 +130,66 @@ protected void testExecute() {
...
+    @Test
+    public void testExecute_addCsrfTokenCookies_shouldAddToResponseAccordingToExistingCsrfToken() {
+
+        ______TS("Test will add CSRF token cookies: No logged in user");
+
+        gaeSimulation.logoutUser();
+
+        MockHttpServletRequest mockRequest = new MockHttpServletRequest(getRequestMethod(),
+                Const.ResourceURIs.URI_PREFIX + getActionUri());
+        mockRequest.addCookie(new Cookie(Const.CsrfConfig.TOKEN_COOKIE_NAME, "some random csrf token"));
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/webapi/GetAuthInfoActionTest.java]: I believe u have done some research on CRSF token.

Here are some typical scenario of the CRSF token.

- User visits the website first time -> CSRF token missing -> CSRF mismatched -> regenerate CSRF token
- User visits the website with existing CSRF token -> CSRF token matched -> do nothing
- User visits the website with expired CSRF token -> CSRF token mismatched -> regenerate CSRF token
- User visits the website with fake CSRF token -> CSRF mismatched -> regenerate CSRF token

Therefore, It is not really that CSRF token is "random".

File: src/test/java/teammates/test/cases/webapi/GetAuthInfoActionTest.java [hunk 7]
```diff
@@ -125,9 +130,66 @@ protected void testExecute() {
         assertThrows(UnauthorizedAccessException.class, () -> getAction(new String[] {
                 Const.ParamsNames.USER_ID, "idOfAnotherInstructor",
         }));
+    }
+
+    @Test
+    public void testExecute_addCsrfTokenCookies_shouldAddToResponseAccordingToExistingCsrfToken() {
+
+        ______TS("Test will add CSRF token cookies: No logged in user");
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/webapi/GetAuthInfoActionTest.java]: What does "Test will add CSRF token cookies" really means? Just `No logged in user` is enough to describe the scenario?

File: src/main/java/teammates/ui/webapi/action/JsonResult.java [hunk 8]
```diff
@@ -20,12 +21,19 @@
...
 
     public JsonResult(ApiOutput output) {
         super(HttpStatus.SC_OK);
         this.output = output;
     }
 
+    public JsonResult(ApiOutput output, Cookie cookie) {
+        super(HttpStatus.SC_OK);
+        this.output = output;
+        this.cookie = cookie;
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/action/JsonResult.java]: The above two line can be simplified to `this(output);`

File: src/main/java/teammates/ui/webapi/action/JsonResult.java [hunk 9]
```diff
@@ -20,12 +21,19 @@
...
 
     private final ApiOutput output;
+    private Cookie cookie;
 
     public JsonResult(ApiOutput output) {
         super(HttpStatus.SC_OK);
         this.output = output;
     }
 
+    public JsonResult(ApiOutput output, Cookie cookie) {
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/action/JsonResult.java]: I think it is better to accept a list of cookies

File: src/main/java/teammates/ui/webapi/action/JsonResult.java [hunk 10]
```diff
@@ -48,4 +56,11 @@ public void send(HttpServletResponse resp) throws IOException {
         pw.print(JsonUtils.toJson(output));
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/action/JsonResult.java]: You need to send cookie here.

File: src/main/java/teammates/ui/webapi/action/GetAuthInfoAction.java [hunk 11]
```diff
@@ -57,6 +59,9 @@ public ActionResult execute() {
             csrfTokenCookie.setSecure(!Config.isDevServer());
             csrfTokenCookie.setPath("/");
             resp.addCookie(csrfTokenCookie);
+            ArrayList<Cookie> cookieLst = new ArrayList<>();
+            cookieLst.add(csrfTokenCookie);
+            return new JsonResult(output, cookieLst);
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/action/GetAuthInfoAction.java]: There is a faster method to create list either from `guava` or `Arrays`

File: src/main/java/teammates/ui/webapi/action/JsonResult.java [hunk 12]
```diff
@@ -46,6 +54,18 @@ public void send(HttpServletResponse resp) throws IOException {
         resp.setContentType("application/json");
         PrintWriter pw = resp.getWriter();
         pw.print(JsonUtils.toJson(output));
+        if (cookies != null && !cookies.isEmpty()) {
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/action/JsonResult.java]: You can initialize cookies as empty list so you don't need the empty check.

File: src/main/java/teammates/ui/webapi/action/JsonResult.java [hunk 13]
```diff
@@ -46,6 +54,18 @@ public void send(HttpServletResponse resp) throws IOException {
...
+                resp.addCookie(cookie);
+            }
+        }
     }
 
+    public List<Cookie> getCookies() {
+        return cookies;
+    }
+
+    public String getFirstCookieValue() {
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/action/JsonResult.java]: Just getCookies() and return a list?

File: src/test/java/teammates/test/driver/GaeSimulation.java [hunk 14]
```diff
@@ -206,9 +207,11 @@ public void loginAsStudentInstructor(String userId) {
      * @param method The request method
      * @param body The request body
      * @param parts The request parts
+     * @param cookieValue The request's cookie value
      * @param params Parameters that appear in a HttpServletRequest received by the app
      */
-    public Action getActionObject(String uri, String method, String body, Map<String, Part> parts, String... params) {
+    public Action getActionObject(String uri, String method, String body, Map<String, Part> parts,
+                                  String cookieValue, String... params) {
```
Comment (xpdavid) [inline src/test/java/teammates/test/driver/GaeSimulation.java]: `String cookieValues` should be `List<Cookie> cookies`?

The method is supposed to be generic. There are many cookies besides CSRF.

File: src/main/java/teammates/ui/webapi/action/GetAuthInfoAction.java [hunk 15]
```diff
@@ -57,6 +60,8 @@ public ActionResult execute() {
             csrfTokenCookie.setSecure(!Config.isDevServer());
             csrfTokenCookie.setPath("/");
             resp.addCookie(csrfTokenCookie);
+            ArrayList<Cookie> cookieLst = new ArrayList<>(Arrays.asList(csrfTokenCookie));
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/action/GetAuthInfoAction.java]:
```suggestion
            List<Cookie> cookieLst = Arrays.asList(csrfTokenCookie);
```

File: src/main/java/teammates/ui/webapi/action/JsonResult.java [hunk 16]
```diff
@@ -46,6 +56,15 @@ public void send(HttpServletResponse resp) throws IOException {
         resp.setContentType("application/json");
         PrintWriter pw = resp.getWriter();
         pw.print(JsonUtils.toJson(output));
+        if (cookies != null) {
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/action/JsonResult.java]: You don't need the null-check?

File: src/test/java/teammates/test/cases/webapi/BaseActionTest.java [hunk 17]
```diff
@@ -84,7 +85,14 @@ protected T getActionWithParts(String key, String filePath, String... params) th
         Map<String, Part> parts = new HashMap<>();
         parts.put(key, new MockPart(filePath));
 
-        return getAction(null, parts, params);
+        return getAction(null, parts, null, params);
+    }
+
+    /**
+     * Gets an action with request multipart config.
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/webapi/BaseActionTest.java]: Wrong Javadoc?

File: src/main/java/teammates/ui/webapi/action/GetAuthInfoAction.java [hunk 18]
```diff
@@ -57,6 +60,8 @@ public ActionResult execute() {
             csrfTokenCookie.setSecure(!Config.isDevServer());
             csrfTokenCookie.setPath("/");
             resp.addCookie(csrfTokenCookie);
+            List<Cookie> cookieLst = Arrays.asList(csrfTokenCookie);
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/action/GetAuthInfoAction.java]: You don't need the abbreviation? Just `cookieList` is fine?

File: src/test/java/teammates/test/cases/webapi/GetAuthInfoActionTest.java [hunk 19]
```diff
@@ -125,9 +132,56 @@ protected void testExecute() {
...
                 Const.ParamsNames.USER_ID, "idOfAnotherInstructor",
         }));
+    }
+
+    @Test
+    public void testExecute_addCsrfTokenCookies_shouldAddToResponseAccordingToExistingCsrfToken() {
+
+        MockHttpServletRequest mockRequest = new MockHttpServletRequest(getRequestMethod(),
+                Const.ResourceURIs.URI_PREFIX + getActionUri());
+        String expectedCsrfToken = StringHelper.encrypt(mockRequest.getSession().getId());
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/webapi/GetAuthInfoActionTest.java]: You can hard-code the session id `1234`?

File: src/test/java/teammates/test/driver/GaeSimulation.java [hunk 20]
```diff
@@ -226,6 +229,11 @@ public Action getActionObject(String uri, String method, String body, Map<String
                     }
                 });
             }
+            if (cookies != null) {
+                for (Cookie cookie : cookies) {
+                    req.addCookie(new Cookie(Const.CsrfConfig.TOKEN_COOKIE_NAME, cookie.getValue()));
```
Comment (xpdavid) [inline src/test/java/teammates/test/driver/GaeSimulation.java]: Shouldn't be
```suggestion
                    req.addCookie(cookie);
```
?

File: src/main/java/teammates/ui/webapi/action/JsonResult.java [hunk 21]
```diff
@@ -46,6 +57,13 @@ public void send(HttpServletResponse resp) throws IOException {
         resp.setContentType("application/json");
         PrintWriter pw = resp.getWriter();
         pw.print(JsonUtils.toJson(output));
+        for (Cookie cookie : cookies) {
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/action/JsonResult.java]: Just a minor issue.

The sending of cookie should before the sending of body. In HTTP, cookie is before the body. Not sure the wrong order will cause problems or not.

File: src/main/java/teammates/ui/webapi/action/GetAuthInfoAction.java [hunk 22]
```diff
@@ -57,6 +60,8 @@ public ActionResult execute() {
             csrfTokenCookie.setSecure(!Config.isDevServer());
             csrfTokenCookie.setPath("/");
             resp.addCookie(csrfTokenCookie);
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/action/GetAuthInfoAction.java]: You can remove this line (and the `resp` in the `Action`)?

**Pull request merged**